go-controller/pkg/node/controllers/egressip/egressip.go

```diff
@@ -0,0 +1,721 @@
+			if linkAddress.Label != linkmanager.GetAssignedAddressLabel(linkName) {
+				continue
+			}
+			isValidEIP = false
+			for _, validAddress := range validAddresses {
+				if validAddress.IP.Equal(linkAddress.IP) {
+					isValidEIP = true
+					break
+				}
+			}
+
+			if !isValidEIP {
+				if err := c.linkManager.DelAddress(linkAddress); err != nil {
+					klog.Errorf("Failed to delete egress IP %s to link manager: %v", linkAddress.IP.String(), err)
+				}
+			}
+		}
+	}
+}
+
+func getRouteTableID(linkIndex int) int {
+	return linkIndex + routeTableStart
+}
+
+func (c *Controller) GetLinkContainingIPAndValidate(ip net.IP, expectedNetwork string) (bool, netlink.Link, error) {
+	prefixLinks := map[string]netlink.Link{} // key is network CIDR
+	prefixes := make([]netip.Prefix, 0)
+	links, err := netlink.LinkList()
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to list links: %v", err)
+	}
+	for _, link := range links {
+		link := link
+		linkPrefixes, err := linkmanager.GetExternallyAvailableNetipAddresses(link, c.v4, c.v6, skipBridges)
+		if err != nil {
+			klog.Errorf("Failed to get address from link %s: %v", link.Attrs().Name, err)
+			continue
+		}
+		prefixes = append(prefixes, linkPrefixes...)
+		// create lookup table for later retrieval
+		for _, prefixFound := range linkPrefixes {
+			_, ipNet, err := net.ParseCIDR(prefixFound.String())
+			if err != nil {
+				klog.Errorf("Egress IP: skipping prefix %q due to parsing CIDR error: %v", prefixFound.String(), err)
+				continue
+			}
+			prefixLinks[ipNet.String()] = link
+		}
+	}
+	lpmTree := cidrtree.New(prefixes...)
+	addr, err := netip.ParseAddr(ip.String())
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to convert IP %s to netip addr: %v", ip.String(), err)
+	}
+	network, found := lpmTree.Lookup(addr)
+	if !found {
+		return false, nil, nil
+	}
+	if network.String() != expectedNetwork {
+		return false, nil, fmt.Errorf("EgressIP %s is was assigned to network %s but we got network %s. Mismatch therefore not assigning",
+			ip.String(), expectedNetwork, network.String())
+	}
+	return true, prefixLinks[network.String()], nil
+}
+
+func getNetworkInfoFromIP(ipStr string) (net.IP, *net.IPNet, bool, error) {
+	var err error
+	var ip net.IP
+	var ipNet *net.IPNet
+	var v6 bool
+	ip = net.ParseIP(ipStr)
+	if utilnet.IsIPv6(ip) {
+		v6 = true
+		ip, ipNet, err = net.ParseCIDR(fmt.Sprintf("%s/128", ipStr))
+	} else {
+		ip, ipNet, err = net.ParseCIDR(fmt.Sprintf("%s/32", ipStr))
+	}
+	return ip, ipNet, v6, err
+}
+
+func filterRouteByLinkTable(linkIndex, tableID int) (*netlink.Route, uint64) {
+	return &netlink.Route{
+			LinkIndex: linkIndex,
+			Table:     tableID,
+		},
+		netlink.RT_FILTER_OIF | netlink.RT_FILTER_TABLE
+}
+
+func filterRuleByPriority(priority int) (*netlink.Rule, uint64) {
+	return &netlink.Rule{
+			Priority: priority,
+		},
+		netlink.RT_FILTER_PRIORITY
+}
+
+func areNetlinkRulesEqual(r1, r2 netlink.Rule) bool {
```

Comment:
Can you use `reflect.DeepEqual` instead?

Reply:
done